python/pfs/drp/stella/readLineList.py

```diff
@@ -47,37 +85,86 @@ def run(self, detectorMap=None, metadata=None):
+
+        if len(self.config.lampElements) > 0:
+            lampElementSet = set(self.config.lampElements)
+        else:
+            if lampInfoFromMetaData:
+                lampElementSet = lampInfoFromMetaData.lampElementSet
+
         lines = ReferenceLineSet.empty()
-        for filename in self.config.lineListFiles:
-            lines.extend(ReferenceLineSet.fromLineList(filename))
-        if self.config.restrictByLamps:
-            lamps = self.getLamps(metadata)
-            lines = self.filterByLamps(lines, lamps)
+        for ll in lightSources:
+            if ll not in self.config.lightSourceMap:
+                raise RuntimeError(f'Light source {ll} '
+                                   f'is not listed in lightSourceMap {self.config.lightSourceMap}')
+
+            filename = self.config.lightSourceMap[ll]
+            if filename is not None:
+                lines.extend(ReferenceLineSet.fromLineList(filename))
+
+        lines = self.filterByLampElements(lines, lampElementSet)
         lines = self.filterByIntensity(lines)
         lines = self.filterByWavelength(lines, detectorMap)
+        self.rejectDuplicates(lines)
         lines.applyExclusionZone(self.config.exclusionRadius)
         return lines
 
-    def filterByLamps(self, lines, lamps):
-        """Filter the line list by which lamps are active
+    def rejectDuplicates(self, lines):
+        """Reject duplicate lines by wavelength from the line list.
+
+        This is performed in-place.
+
+        This may occur if multiple input sources may share
+        lines with the same wavelength.
 
         Parameters
         ----------
         lines : `pfs.drp.stella.ReferenceLineSet`
             List of reference lines.
-        lamps : ``list` of `str`
-            The list of lamps.
+        """
+        _, indices = np.unique(lines.wavelength, return_index=True)
+        reject = np.full(len(lines), True, dtype=bool)
+        reject[indices] = False
+        lines.status[reject] |= ReferenceLineStatus.BLEND
+
+    def filterByLampElements(self, lines, lampElementSet):
+        """Filter the line list by the elements or species in the active lamps
+
+        Parameters
+        ----------
+        lines : `pfs.drp.stella.ReferenceLineSet`
+            List of reference lines.
+        elementSet : ``set` of `str`
+            The set of elements or species in the active lamps.
 
         Returns
         -------
         filtered : `pfs.drp.stella.ReferenceLineSet`
-            Filtered list of reference lines.
+            Filtered list of reference lines, or input lines if ``elements`` is Falsey.
         """
-        select = np.zeros(len(lines), dtype=bool)
-        for desc in lamps:
-            select |= [ll.description.startswith(desc) for ll in lines]
-        self.log.info("Filtered line lists for lamps: %s", ",".join(sorted(lamps)))
-        return lines[select]
+        if lampElementSet is None or len(lampElementSet) == 0:
```

Comment:
The possibility of `lampElementSet` being `None` isn't mentioned in the docstring.